tests/test_androidmodule_ctypes_finder.py

```diff
@@ -0,0 +1,131 @@
+
+sys.modules["jnius"] = MagicMock()  # fake jnius used later
+
+
+# Import the tested android._ctypes_library_finder module,
+# making sure android._android won't crash us!
+# (since android._android is android-only / not compilable on desktop)
+android_module_folder = os.path.abspath(os.path.join(
+    os.path.dirname(__file__),
+    "..", "pythonforandroid", "recipes", "android", "src"
+))
+sys.path.insert(0, android_module_folder)
+sys.modules['android._android'] = MagicMock()
+import android._ctypes_library_finder
+sys.path.remove(android_module_folder)
+
+
+def _get_autoclass_with_fake_activity(returning_test_path="/testpath",
+                                      return_as_service=False):
+    """
+    Helper functions to mock autoclass() from jnius.
+
+    This will return a mocked activity "org.kivy.android.PythonActivity"
+    which supports .getApplicationContext().getPackagemanager(),
+    with that returned object supporting .getApplicationInfo()
+    for PackageManager's GET_SHARED_LIBRARY_FILES info.
+
+    (That is the functionality the android._ctypes_library_finder module
+     uses to scan the application-specific library directories)
+    """
+
+    # This is a closure to bind the value of "returning_test_path":
+    def autoclass_returning_graphical_activity(name):
+        activity = MagicMock()
+        activity.getApplicationContext().getPackageName = (
+             lambda: "com.example.testpackage"
+        )
+        activity.getApplicationContext().getPackageManager().\
+            getApplicationInfo().nativeLibraryDir = returning_test_path
+        activity_class = MagicMock()
+
+        if name == "org.kivy.android.PythonActivity" and \
+                not return_as_service:
+            activity_class.mService = None
+            activity_class.mActivity = activity
+            return activity_class
+        elif name == "org.kivy.android.PythonService" and return_as_service:
+            activity_class.mActivity = None
+            activity_class.mService = activity
+            return activity_class
+        elif name == "android.content.pm.PackageManager":
+            manager_class = MagicMock()
+            manager_class.GET_SHARED_LIBRARY_FILES = 1024
+            return manager_class
+        return None
+
+    return autoclass_returning_graphical_activity
+
+
+def test_get_activity_lib_dir():
+    import jnius  # should get us our fake module
+
+    jnius.autoclass = MagicMock()
+    jnius.autoclass.return_value = None
+    assert android._ctypes_library_finder.get_activity_lib_dir(
+        "JavaClass"
+    ) is None
+    assert mock.call("JavaClass") in jnius.autoclass.call_args_list
+
+    jnius.autoclass = _get_autoclass_with_fake_activity("/testpath")
+    assert android._ctypes_library_finder.get_activity_lib_dir(
+        "org.kivy.android.PythonActivity"
+    ) == "/testpath"
+    assert android._ctypes_library_finder.get_activity_lib_dir(
+        "random.invalid.name"
+    ) is None
```

Comment:
This is a good example showing that this test is not testing your `get_activity_lib_dir()` code but it's testing the test util.
And this is exactly what I want to avoid with having code branching via utils functions in tests.
Do you see what I mean?

Reply:
Basically the fact that `autoclass` returns `None` on non-existing java class is not a feature of `get_activity_lib_dir` but it's a feature of `autoclass`, however our point is not to unit test `autoclass`, we assume `autoclass` is acting correctly. That's why a test like I wrote as a quick example `test_get_activity_lib_dir_autoclass_init_none()` would make more sense. Do you reckon that the way your write it makes it *feels* like it's a feature of `get_activity_lib_dir()` that it would behave differently depending on `activity_name`?

Reply:
sure. it's revamped now to check the call args instead :+1: